.github/workflows/pull-request.yml

```diff
@@ -0,0 +1,71 @@
+name: pull request
+
+on: [pull_request]
+
+jobs:
+  build:
+    name: build
+    runs-on: ubuntu-latest
+
+    outputs:
+      cache_key: ${{ steps.cache_key.outputs.cache_key }}
+
+    steps:
+      - uses: actions/checkout@v3
+      - id: cache_key
+        run: echo "cache_key=${{ hashFiles('**/**') }}-v1" >> "$GITHUB_OUTPUT"
+      - uses: actions/cache@v3
+        id: cache
+        with:
+          path: .
+          key: ${{ steps.cache_key.outputs.cache_key }}
+      - if: steps.cache.outputs.hit != true
+        uses: actions/setup-node@v3
+        with:
+          node-version: 18
+      - if: steps.cache.outputs.hit != true
+        run: npm install
+      - if: steps.cache.outputs.hit != true
+        run: npm run build
+
+  validate_html:
+    needs: [build]
+    name: validate html
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/cache/restore@v3
+        with:
+          key: ${{ needs.build.outputs.cache_key }}
+          path: .
+      - uses: actions/setup-node@v3
```

Comment:
i don't do GitHub Actions enough to remember their idiosyncracies in-between reviews -- are these three jobs re-setting up node? since they require the build step, can the repeated uses of `actions/cache/restore@v3` and `actions/setup-node@v3` go away?

Reply:
I wish! Each "job" runs in its own container, so they start fresh. To get rid of the duplicated steps, all the jobs could be merged into one, but I chose to split them up so they could run in parallel. Which turns out to be a particularly useful optimization since the tests themselves take a few seconds each. The time spent creating, uploading, and restoring the cache might very well be crushing any time saved with parallelized tests. 🤔

Reply:
hahahaah wheeeeee

Reply:
Yes. It might be worth changing them in another PR? The problem with having them all in a single job is that they show up as a single item in the list of GitHub checks, which makes it harder to understand what went wrong when they fail. 😫